modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/ViewsIntegrationTest.java

```diff
@@ -0,0 +1,432 @@
+        sql("create or replace view cache1.my_view as select * from public.my_table");
+        sql("create or replace view my_schema.my_view as select * from public.my_table");
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+        assertViewsCount("CACHE1", "MY_VIEW", 1);
+        assertViewsCount("MY_SCHEMA", "MY_VIEW", 1);
+
+        client.destroyCache("CACHE1");
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+        assertViewsCount("CACHE1", "MY_VIEW", 0);
+        assertViewsCount("MY_SCHEMA", "MY_VIEW", 1);
+
+        client.destroyCache("CACHE2");
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+        assertViewsCount("CACHE1", "MY_VIEW", 0);
+        assertViewsCount("MY_SCHEMA", "MY_VIEW", 1);
+
+        client.destroyCache("CACHE3");
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+        assertViewsCount("CACHE1", "MY_VIEW", 0);
+        assertViewsCount("MY_SCHEMA", "MY_VIEW", 1);
+
+        client.destroyCache("CACHE4");
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+        assertViewsCount("CACHE1", "MY_VIEW", 0);
+        assertViewsCount("MY_SCHEMA", "MY_VIEW", 0);
+
+        client.getOrCreateCache(new CacheConfiguration<>("CACHE1")
+            .setIndexedTypes(Integer.class, Integer.class));
+
+        client.getOrCreateCache(new CacheConfiguration<>("CACHE3")
+            .setIndexedTypes(Integer.class, Integer.class)
+            .setSqlSchema("MY_SCHEMA")
+        );
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+        assertViewsCount("CACHE1", "MY_VIEW", 0);
+        assertViewsCount("MY_SCHEMA", "MY_VIEW", 0);
+    }
+
+    /**
+     * Tests views on not existing schema.
+     */
+    @Test
+    public void testNotExistingSchema() throws Exception {
+        initGrids(3);
+        initTable(3);
+
+        assertThrows("CREATE VIEW my_schema.my_view AS SELECT * FROM public.my_table", IgniteSQLException.class,
+            "MY_SCHEMA");
+    }
+
+    /**
+     * Tests view after in-memory grid restart.
+     */
+    @Test
+    public void testInMemory() throws Exception {
+        initGrids(3);
+        initTable(3);
+
+        sql("create or replace view my_view as select * from my_table");
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+
+        stopAllGrids();
+
+        initGrids(3);
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 0);
+    }
+
+    /**
+     * Tests view after persistent grid restart.
+     */
+    @Test
+    public void testPersistance() throws Exception {
+        persistenceEnabled = true;
+
+        initGrids(3);
+        initTable(3);
+
+        sql("create or replace view my_view as select * from my_table");
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+
+        stopAllGrids();
+
+        initGrids(3);
+
+        assertViewsCount("PUBLIC", "MY_VIEW", 1);
+    }
+
+    /**
+     * Tests views information distribution on node join.
+     */
+    @Test
+    public void testNodeJoin() throws Exception {
```

Comment:
What if only non-baseline node left? Will a view exist?

Reply:
Distributed metastorage is stored on non-baseline nodes too (AFAIK)